ddd/ddd-02-conn-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,167 @@
+       a poll caused by WANT_WRITE (more likely) / wants to listen for POLLOUT even
+       during a poll caused by WANT_READ (less likely but possible, e.g. frames
+       transmitted on a timer)?
+
+    PROPOSED SOLUTION: POLLIN should always be listened for.
+
+    CHANGE: Applications should always listen for POLLIN.
+
+    Q. What if the QUIC stack starts to need to read during WANT_WRITE polling?
+    A. Possibly the application should simply always poll for POLLIN.
+       We probably want the QUIC state machine to always handle incoming frames
+       immediately even if TX queues are full so this would make sense.
+
+    PROPOSED SOLUTION: POLLIN should always be listened for.
+
+    CHANGE: If applications do not already test POLLIN during WANT_WRITE they
+        will need to be changed to do so. Arguably they should already do this to
+        handle TLS close notifications etc.
+
+    Q. What if the QUIC stack starts to need to write during WANT_READ polling?
+    A. The QUIC stack should be able to predict at what point in the future
+       it will want to do a spontaneous write so this can be handled by whatever
+       solution is conceived for 'timer issues' below.
+
+    PROPOSED SOLUTION: This is handled by whatever solution we choose for our timer
+        issue. After timer events are handled SSL_read/SSL_write should be
+        tried again which results in a new SSL_ERROR_WANT_READ/WRITE return
+        code. If libssl wants to be notified for both POLLIN and POLLOUT
+        it can return SSL_ERROR_WANT_WRITE due to the decision that POLLIN
+        should always be handled above.
+
+        Applications will need to ensure they retry SSL_read/SSL_write
+        after a timer event (i.e. timer events effectively imply POLLIN|POLLOUT
+        from an application perspective.)
+
+        (Should we also offer the abillity to get a SSL_ERROR_WANT_READ/WRITE
+         as a result of the BIO_pump call?)
+
+  - Timer issues:
+
+    - The QUIC stack may need to randomly send frames (e.g. keepalives)
+      even when the application is not calling SSL_read/SSL_write for a long
+      time (e.g. an hour).
+
+    - So arrangements must be made for some call to libssl to be made
+      periodically.
+
+      PROPOSED SOLUTION: A call, BIO_pump (/SSL_pump) is introduced which can be used
+      when an application wishes to allow libssl to process events without
+      calling read/write functions.
+
+      CHANGE: See below.
+
+    - It is possible this can be handled automatically on platforms where
+      threads can be used but this cannot be assumed on all platforms.
+
+    - Worst case, provide a "get wakeup time" function along with the "get FD"
+      function which indicates the deadline by which libssl should next be
+      called.
+
+      PROPOSED SOLUTION: A function BIO_get_timeout is introduced which returns
+      when libssl next wants to be called.
```

Comment:
`BIO_get_timeout` doesn't seem like a good fit. I would expect this to be an "SSL" function not a "BIO" function. The DTLS code already has `DTLSv1_get_timeout()` in openssl/openssl#17184 I proposed `SSL_get_next_tick()`.

Reply:
I'd be totally happy with either of these.

Where a `BIO_*` function is mentioned in this proposal generally the implication is it just calls the equivalent `SSL_*` function for `BIO_f_ssl` and is otherwise unimplemented, it's just a helper method. But we can also make the client do it itself and call the `SSL_*` function directly, it doesn't really matter.

Maybe we can deprecate `DTLSv1_get_timeout` and replace it with `SSL_get_next_tick` in future.

Reply:
> Whether it warrants deprecation, I'm not sure.

Unless the name is confusing by itself (i.e., even if used for DTLSv1) I would not deprecate it. I'd just keep both names.